core/primitives/src/congestion_info.rs

```diff
@@ -318,6 +320,50 @@ impl CongestionInfo {
     }
 }
 
+// TODO remove default
+#[derive(Clone, Debug, Default)]
+pub struct BlockCongestionInfo {
+    shards_congestion_info: BTreeMap<ShardId, ExtendedCongestionInfo>,
+}
+
+impl BlockCongestionInfo {
+    pub fn new(shards_congestion_info: BTreeMap<ShardId, ExtendedCongestionInfo>) -> Self {
+        Self { shards_congestion_info }
+    }
+
+    pub fn iter(&self) -> std::collections::btree_map::Iter<'_, u64, ExtendedCongestionInfo> {
+        self.shards_congestion_info.iter()
+    }
+
+    pub fn all_shards(&self) -> Vec<ShardId> {
+        self.shards_congestion_info.keys().copied().collect()
+    }
+
+    pub fn get(&self, shard_id: &ShardId) -> Option<&ExtendedCongestionInfo> {
+        self.shards_congestion_info.get(shard_id)
+    }
+
+    pub fn get_mut(&mut self, shard_id: &ShardId) -> Option<&mut ExtendedCongestionInfo> {
+        self.shards_congestion_info.get_mut(shard_id)
+    }
+
+    pub fn insert(
+        &mut self,
+        shard_id: ShardId,
+        value: ExtendedCongestionInfo,
+    ) -> Option<ExtendedCongestionInfo> {
+        self.shards_congestion_info.insert(shard_id, value)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.shards_congestion_info.is_empty()
+    }
+
+    pub fn clear(&mut self) -> () {
+        self.shards_congestion_info.clear()
+    }
```

Comment:
nit: this feels like C++ style mutability, why not overwrite the whole variable?

Reply:
cool observation, let me try and see if it's cleaner and rustier